### ISSUE TITLE
Implement U2F

### ISSUE THREAD

**Issue by preisi** [sddm/sddm#784]
Hello,
as part of a university linux project we (a team of 2 students) want to contribute to SDDM.

This goal is rather self-explanatory:
We want to add U2F functionality for SDDM.

Please note that the other 3 issues opened by me will have a higher priority for us because these contributions are part of a university project.

What do you think about this planned functionality?

Cheers,
Thomas and Milan

P.S.: Of course we will develop this feature in the open on GitHub and give updates here.

**Comment by m1st0** [comment]
After setting up pam_u2f and appropriate keys in a configuration directory of my choosing in Ubuntu Zesty, I got SDDM to work with U2F.  UI feedback and interaction would be a future goal of mine, however the key operates correctly either as a 2FA device or a password-less login option.  The successful point I found it working at is adding to **/etc/pam.d/sddm** the following:
```
#%PAM-1.0

# Block login if they are globally disabled
auth    requisite       pam_nologin.so
auth    required        pam_succeed_if.so user != root quiet_success

# U2F options
auth sufficient pam_u2f.so authfile=/home/my_home_directory/.config/my_purchased_u2f_key/u2f_keys
```

For testing I used _auth sufficient_.  I was able to use directly by clicking login and pushing the button on the key and also as a 2FA option after password entry and clicking login. If the key is not active, with the **auth sufficient** option I can login directly with a password.  With **auth require** I believe it would only work as a 2FA device once the password is provided and login is clicked.  SDDM appropriately awaited for the key to be clicked if it is active.  Is the next step is to add UI/UX to work with these options alongside PAM as well?

**Comment by tacofrog2** [comment]
Like @m1st0 I got SDDM to work with pam_u2f.so, where after the password is entered, SDDM waits for the key to be entered. 
I cannot get SDDM to work with the pam_openotp.so module. The lock screen and login screen just gives a "login failed" error without asking for the key. So it seems that U2F is supported but not other two-factor methods.

**Comment by julianbrost** [comment]
Yes, pam_u2f.so works with any login manager it does not interact with it but only directly with the U2F device. The user experience is far from great as is looks just like the login hangs if you forget to press the button. Also, if there is no U2F device connected, the login will fail instantly. So it would be nice if there was a prompt that asks the user to press the button or connect the device.

**Comment by m1st0** [comment]
On my system it flashes an LED on my U2F device. Yet otherwise with a
proper configuration, I can use either a password and/or U2F device. This
defeats the case of 2FA, but I can cover this situation even if the login
manager's UI/UX isn't ready yet. PAM handles all of the interaction.
